{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA-seq normalizations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The history saving thread hit an unexpected error (DatabaseError('database disk image is malformed')).History will not be written to the database.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_counts = \"/cellar/users/aklie/data/datasets/Hangauer2017_BT474_RNA-seq/results/quant/featureCounts/hangauer.results.counts\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BT474_Parental_rep_1</th>\n",
       "      <th>BT474_Parental_rep_2</th>\n",
       "      <th>BT474_Persister_rep_1</th>\n",
       "      <th>BT474_Persister_rep_2</th>\n",
       "      <th>BT474_Persister_rep_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>48</td>\n",
       "      <td>58</td>\n",
       "      <td>121</td>\n",
       "      <td>114</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   BT474_Parental_rep_1  BT474_Parental_rep_2  \\\n",
       "Geneid                                                          \n",
       "ENSG00000223972.4                     1                     0   \n",
       "ENSG00000227232.4                    48                    58   \n",
       "ENSG00000243485.2                     0                     0   \n",
       "ENSG00000237613.2                     0                     0   \n",
       "ENSG00000268020.2                     0                     0   \n",
       "\n",
       "                   BT474_Persister_rep_1  BT474_Persister_rep_2  \\\n",
       "Geneid                                                            \n",
       "ENSG00000223972.4                      2                      0   \n",
       "ENSG00000227232.4                    121                    114   \n",
       "ENSG00000243485.2                      0                      0   \n",
       "ENSG00000237613.2                      0                      0   \n",
       "ENSG00000268020.2                      0                      0   \n",
       "\n",
       "                   BT474_Persister_rep_3  \n",
       "Geneid                                    \n",
       "ENSG00000223972.4                      1  \n",
       "ENSG00000227232.4                    151  \n",
       "ENSG00000243485.2                      0  \n",
       "ENSG00000237613.2                      0  \n",
       "ENSG00000268020.2                      0  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load counts\n",
    "counts = pd.read_csv(path_counts, comment=\"#\", sep=\"\\t\", index_col=0)\n",
    "gene_info = counts.iloc[:, 0:5]\n",
    "counts = counts.iloc[:, 5:]\n",
    "counts.columns = [x.split(\"/\")[-1].split(\".\")[0] for x in counts.columns]\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CPM Normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rnanorm import CPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_cpm(counts_matrix):\n",
    "    \"\"\"\n",
    "    Normalize an RNA-seq counts matrix using Counts Per Million (CPM).\n",
    "    \n",
    "    Counts Per Million (CPM) is a widely used method in RNA-seq data analysis to normalize \n",
    "    gene expression levels. CPM aims to adjust for differences in sequencing depths across \n",
    "    samples and provide relative expression values on a comparable scale by scaling the raw \n",
    "    read counts of each gene by a sample-specific sequencing depth (total counts) and \n",
    "    multiplying by a scaling factor of one million (to obtain counts per million).\n",
    "\n",
    "    Pros:\n",
    "    - Relatively simple and intuitive to implement.\n",
    "    - Allows for direct comparisons of gene expression levels between samples.\n",
    "    \n",
    "    Cons:\n",
    "    - Does not account for gene length, potentially leading to biases in gene expression estimation.\n",
    "    - May be sensitive to extreme values or outliers.\n",
    "    \n",
    "    Parameters:\n",
    "    counts_matrix (pd.DataFrame): A pandas DataFrame containing raw read counts \n",
    "                                  with rows as genes and columns as samples.\n",
    "    \n",
    "    Returns:\n",
    "    pd.DataFrame: A pandas DataFrame with CPM-normalized expression values.\n",
    "    \n",
    "    Raises:\n",
    "    ValueError: If the input is not a pandas DataFrame.\n",
    "    \"\"\"\n",
    "    \n",
    "    if not isinstance(counts_matrix, pd.DataFrame):\n",
    "        raise ValueError(\"Input must be a pandas DataFrame.\")\n",
    "    \n",
    "    # Calculate total counts per sample\n",
    "    total_counts = counts_matrix.sum(axis=0)\n",
    "    \n",
    "    # CPM normalization\n",
    "    cpm_matrix = counts_matrix.div(total_counts, axis=1) * 1e6\n",
    "    \n",
    "    return cpm_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BT474_Parental_rep_1</th>\n",
       "      <th>BT474_Parental_rep_2</th>\n",
       "      <th>BT474_Persister_rep_1</th>\n",
       "      <th>BT474_Persister_rep_2</th>\n",
       "      <th>BT474_Persister_rep_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>0.008634</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018771</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>0.414419</td>\n",
       "      <td>0.501441</td>\n",
       "      <td>1.135676</td>\n",
       "      <td>0.979852</td>\n",
       "      <td>1.407743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   BT474_Parental_rep_1  BT474_Parental_rep_2  \\\n",
       "Geneid                                                          \n",
       "ENSG00000223972.4              0.008634              0.000000   \n",
       "ENSG00000227232.4              0.414419              0.501441   \n",
       "ENSG00000243485.2              0.000000              0.000000   \n",
       "ENSG00000237613.2              0.000000              0.000000   \n",
       "ENSG00000268020.2              0.000000              0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_1  BT474_Persister_rep_2  \\\n",
       "Geneid                                                            \n",
       "ENSG00000223972.4               0.018771               0.000000   \n",
       "ENSG00000227232.4               1.135676               0.979852   \n",
       "ENSG00000243485.2               0.000000               0.000000   \n",
       "ENSG00000237613.2               0.000000               0.000000   \n",
       "ENSG00000268020.2               0.000000               0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_3  \n",
       "Geneid                                    \n",
       "ENSG00000223972.4               0.009323  \n",
       "ENSG00000227232.4               1.407743  \n",
       "ENSG00000243485.2               0.000000  \n",
       "ENSG00000237613.2               0.000000  \n",
       "ENSG00000268020.2               0.000000  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Normalize counts using CPM\n",
    "cpm_counts = normalize_cpm(counts)\n",
    "cpm_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BT474_Parental_rep_1</th>\n",
       "      <th>BT474_Parental_rep_2</th>\n",
       "      <th>BT474_Persister_rep_1</th>\n",
       "      <th>BT474_Persister_rep_2</th>\n",
       "      <th>BT474_Persister_rep_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>0.008634</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018771</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>0.414419</td>\n",
       "      <td>0.501441</td>\n",
       "      <td>1.135676</td>\n",
       "      <td>0.979852</td>\n",
       "      <td>1.407743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   BT474_Parental_rep_1  BT474_Parental_rep_2  \\\n",
       "ENSG00000223972.4              0.008634              0.000000   \n",
       "ENSG00000227232.4              0.414419              0.501441   \n",
       "ENSG00000243485.2              0.000000              0.000000   \n",
       "ENSG00000237613.2              0.000000              0.000000   \n",
       "ENSG00000268020.2              0.000000              0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_1  BT474_Persister_rep_2  \\\n",
       "ENSG00000223972.4               0.018771               0.000000   \n",
       "ENSG00000227232.4               1.135676               0.979852   \n",
       "ENSG00000243485.2               0.000000               0.000000   \n",
       "ENSG00000237613.2               0.000000               0.000000   \n",
       "ENSG00000268020.2               0.000000               0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_3  \n",
       "ENSG00000223972.4               0.009323  \n",
       "ENSG00000227232.4               1.407743  \n",
       "ENSG00000243485.2               0.000000  \n",
       "ENSG00000237613.2               0.000000  \n",
       "ENSG00000268020.2               0.000000  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# use the rnanorm package to normalize the counts\n",
    "CPM().set_output(transform=\"pandas\").fit_transform(counts.T).T.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TPM Normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rnanorm import TPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_tpm(counts_matrix, gene_lengths):\n",
    "    \"\"\"\n",
    "    Normalize an RNA-seq counts matrix using Transcripts Per Kilobase Million (TPM).\n",
    "    \n",
    "    Transcripts Per Kilobase Million (TPM) is an improvement over RPKM/FPKM. \n",
    "    TPM first normalizes for gene length, then for sequencing depth, making the sum \n",
    "    of all TPMs in each sample identical. This allows for a more accurate comparison \n",
    "    of gene expression between samples.\n",
    "    \n",
    "    Pros:\n",
    "    - More accurate than RPKM/FPKM for comparing gene expression levels between samples.\n",
    "    - Accounts for the total sum of normalized expression levels, allowing for a more balanced comparison.\n",
    "    \n",
    "    Cons:\n",
    "    - Can be affected by highly expressed genes and depends on accurate estimates of gene lengths and accurate read mapping.\n",
    "    - Cannot be used for differential expression analysis.\n",
    "    \n",
    "    Parameters:\n",
    "    counts_matrix (pd.DataFrame): A pandas DataFrame containing raw read counts \n",
    "                                  with rows as genes and columns as samples.\n",
    "    gene_lengths (pd.Series): A pandas Series containing the lengths of genes in kilobases.\n",
    "                              The index should match the row index of the counts_matrix.\n",
    "    \n",
    "    Returns:\n",
    "    pd.DataFrame: A pandas DataFrame with TPM-normalized expression values.\n",
    "    \n",
    "    Raises:\n",
    "    ValueError: If the input counts_matrix is not a pandas DataFrame or if gene_lengths is not a pandas Series.\n",
    "    \"\"\"\n",
    "    \n",
    "    if not isinstance(counts_matrix, pd.DataFrame):\n",
    "        raise ValueError(\"counts_matrix must be a pandas DataFrame.\")\n",
    "    \n",
    "    if not isinstance(gene_lengths, pd.Series):\n",
    "        raise ValueError(\"gene_lengths must be a pandas Series.\")\n",
    "    \n",
    "    if not all(counts_matrix.index == gene_lengths.index):\n",
    "        raise ValueError(\"The index of counts_matrix and gene_lengths must match.\")\n",
    "    \n",
    "    # Normalize by gene length in kilobases\n",
    "    length_normalized = counts_matrix.div(gene_lengths, axis=0)\n",
    "    \n",
    "    # Calculate per-sample scaling factors (total sum of length-normalized counts)\n",
    "    scaling_factors = length_normalized.sum(axis=0)\n",
    "    \n",
    "    # Normalize by scaling factors and multiply by 1e6 to get TPM\n",
    "    tpm_matrix = length_normalized.div(scaling_factors, axis=1) * 1e6\n",
    "    \n",
    "    return tpm_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Geneid\n",
       "ENSG00000223972.4    1756\n",
       "ENSG00000227232.4    2073\n",
       "ENSG00000243485.2    1021\n",
       "ENSG00000237613.2    1219\n",
       "ENSG00000268020.2     947\n",
       "Name: Length, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get gene lengths\n",
    "gene_lengths = gene_info[\"Length\"]\n",
    "gene_lengths.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BT474_Parental_rep_1</th>\n",
       "      <th>BT474_Parental_rep_2</th>\n",
       "      <th>BT474_Persister_rep_1</th>\n",
       "      <th>BT474_Persister_rep_2</th>\n",
       "      <th>BT474_Persister_rep_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>0.015742</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034341</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>0.640070</td>\n",
       "      <td>0.784764</td>\n",
       "      <td>1.759937</td>\n",
       "      <td>1.536768</td>\n",
       "      <td>2.206111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   BT474_Parental_rep_1  BT474_Parental_rep_2  \\\n",
       "Geneid                                                          \n",
       "ENSG00000223972.4              0.015742              0.000000   \n",
       "ENSG00000227232.4              0.640070              0.784764   \n",
       "ENSG00000243485.2              0.000000              0.000000   \n",
       "ENSG00000237613.2              0.000000              0.000000   \n",
       "ENSG00000268020.2              0.000000              0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_1  BT474_Persister_rep_2  \\\n",
       "Geneid                                                            \n",
       "ENSG00000223972.4               0.034341               0.000000   \n",
       "ENSG00000227232.4               1.759937               1.536768   \n",
       "ENSG00000243485.2               0.000000               0.000000   \n",
       "ENSG00000237613.2               0.000000               0.000000   \n",
       "ENSG00000268020.2               0.000000               0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_3  \n",
       "Geneid                                    \n",
       "ENSG00000223972.4               0.017247  \n",
       "ENSG00000227232.4               2.206111  \n",
       "ENSG00000243485.2               0.000000  \n",
       "ENSG00000237613.2               0.000000  \n",
       "ENSG00000268020.2               0.000000  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Normalize counts using TPM\n",
    "tpm_counts = normalize_tpm(counts, gene_lengths)\n",
    "tpm_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BT474_Parental_rep_1</th>\n",
       "      <th>BT474_Parental_rep_2</th>\n",
       "      <th>BT474_Persister_rep_1</th>\n",
       "      <th>BT474_Persister_rep_2</th>\n",
       "      <th>BT474_Persister_rep_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>0.015742</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034341</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>0.640070</td>\n",
       "      <td>0.784764</td>\n",
       "      <td>1.759937</td>\n",
       "      <td>1.536768</td>\n",
       "      <td>2.206111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   BT474_Parental_rep_1  BT474_Parental_rep_2  \\\n",
       "ENSG00000223972.4              0.015742              0.000000   \n",
       "ENSG00000227232.4              0.640070              0.784764   \n",
       "ENSG00000243485.2              0.000000              0.000000   \n",
       "ENSG00000237613.2              0.000000              0.000000   \n",
       "ENSG00000268020.2              0.000000              0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_1  BT474_Persister_rep_2  \\\n",
       "ENSG00000223972.4               0.034341               0.000000   \n",
       "ENSG00000227232.4               1.759937               1.536768   \n",
       "ENSG00000243485.2               0.000000               0.000000   \n",
       "ENSG00000237613.2               0.000000               0.000000   \n",
       "ENSG00000268020.2               0.000000               0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_3  \n",
       "ENSG00000223972.4               0.017247  \n",
       "ENSG00000227232.4               2.206111  \n",
       "ENSG00000243485.2               0.000000  \n",
       "ENSG00000237613.2               0.000000  \n",
       "ENSG00000268020.2               0.000000  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# use the rnanorm package to normalize the counts\n",
    "TPM(gene_lengths=gene_lengths).set_output(transform=\"pandas\").fit_transform(counts.T).T.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RPKM/FPKM Normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rnanorm import FPKM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def normalize_rpkm(counts_matrix, gene_lengths):\n",
    "    \"\"\"\n",
    "    Normalize an RNA-seq counts matrix using Reads Per Kilobase of transcript per Million mapped reads (RPKM).\n",
    "    \n",
    "    RPKM (Reads Per Kilobase of transcript per Million mapped reads) and FPKM (Fragments Per Kilobase \n",
    "    of transcript per Million mapped reads) normalize for both the length of the gene and the total \n",
    "    number of reads (i.e., the library size), making expression level comparisons between genes \n",
    "    in the same sample possible.\n",
    "    \n",
    "    Pros:\n",
    "    - Widely used and established/incorporated in several software tools.\n",
    "    - Makes it possible to compare gene expression levels within the same sample and between different samples.\n",
    "    \n",
    "    Cons:\n",
    "    - Assumes that the total number of reads is the same across all samples, which isn't always accurate, \n",
    "      particularly when comparing different conditions or tissues.\n",
    "    - Can be biased by highly expressed genes or transcripts, making it less reliable for datasets \n",
    "      with a high level of expression variation.\n",
    "    - Like TPM, it cannot be used for differential expression analysis.\n",
    "    \n",
    "    Parameters:\n",
    "    counts_matrix (pd.DataFrame): A pandas DataFrame containing raw read counts \n",
    "                                  with rows as genes and columns as samples.\n",
    "    gene_lengths (pd.Series): A pandas Series containing the lengths of genes in kilobases.\n",
    "                              The index should match the row index of the counts_matrix.\n",
    "    \n",
    "    Returns:\n",
    "    pd.DataFrame: A pandas DataFrame with RPKM-normalized expression values.\n",
    "    \n",
    "    Raises:\n",
    "    ValueError: If the input counts_matrix is not a pandas DataFrame or if gene_lengths is not a pandas Series.\n",
    "    \"\"\"\n",
    "    \n",
    "    if not isinstance(counts_matrix, pd.DataFrame):\n",
    "        raise ValueError(\"counts_matrix must be a pandas DataFrame.\")\n",
    "    \n",
    "    if not isinstance(gene_lengths, pd.Series):\n",
    "        raise ValueError(\"gene_lengths must be a pandas Series.\")\n",
    "    \n",
    "    if not all(counts_matrix.index == gene_lengths.index):\n",
    "        raise ValueError(\"The index of counts_matrix and gene_lengths must match.\")\n",
    "    \n",
    "    # Divide by both gene length and total counts, then multiply by 1e9 to get RPKM\n",
    "    rpkm_matrix = counts_matrix.div(gene_lengths, axis=0).div(counts_matrix.sum(axis=0), axis=1) * 1e9\n",
    "    \n",
    "    return rpkm_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BT474_Parental_rep_1</th>\n",
       "      <th>BT474_Parental_rep_2</th>\n",
       "      <th>BT474_Persister_rep_1</th>\n",
       "      <th>BT474_Persister_rep_2</th>\n",
       "      <th>BT474_Persister_rep_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>0.004917</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010690</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>0.199913</td>\n",
       "      <td>0.241892</td>\n",
       "      <td>0.547842</td>\n",
       "      <td>0.472673</td>\n",
       "      <td>0.679085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   BT474_Parental_rep_1  BT474_Parental_rep_2  \\\n",
       "Geneid                                                          \n",
       "ENSG00000223972.4              0.004917              0.000000   \n",
       "ENSG00000227232.4              0.199913              0.241892   \n",
       "ENSG00000243485.2              0.000000              0.000000   \n",
       "ENSG00000237613.2              0.000000              0.000000   \n",
       "ENSG00000268020.2              0.000000              0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_1  BT474_Persister_rep_2  \\\n",
       "Geneid                                                            \n",
       "ENSG00000223972.4               0.010690               0.000000   \n",
       "ENSG00000227232.4               0.547842               0.472673   \n",
       "ENSG00000243485.2               0.000000               0.000000   \n",
       "ENSG00000237613.2               0.000000               0.000000   \n",
       "ENSG00000268020.2               0.000000               0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_3  \n",
       "Geneid                                    \n",
       "ENSG00000223972.4               0.005309  \n",
       "ENSG00000227232.4               0.679085  \n",
       "ENSG00000243485.2               0.000000  \n",
       "ENSG00000237613.2               0.000000  \n",
       "ENSG00000268020.2               0.000000  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Normalize counts using RPKM\n",
    "rpkm_counts = normalize_rpkm(counts, gene_lengths)\n",
    "rpkm_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BT474_Parental_rep_1</th>\n",
       "      <th>BT474_Parental_rep_2</th>\n",
       "      <th>BT474_Persister_rep_1</th>\n",
       "      <th>BT474_Persister_rep_2</th>\n",
       "      <th>BT474_Persister_rep_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>0.004917</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010690</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>0.199913</td>\n",
       "      <td>0.241892</td>\n",
       "      <td>0.547842</td>\n",
       "      <td>0.472673</td>\n",
       "      <td>0.679085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   BT474_Parental_rep_1  BT474_Parental_rep_2  \\\n",
       "ENSG00000223972.4              0.004917              0.000000   \n",
       "ENSG00000227232.4              0.199913              0.241892   \n",
       "ENSG00000243485.2              0.000000              0.000000   \n",
       "ENSG00000237613.2              0.000000              0.000000   \n",
       "ENSG00000268020.2              0.000000              0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_1  BT474_Persister_rep_2  \\\n",
       "ENSG00000223972.4               0.010690               0.000000   \n",
       "ENSG00000227232.4               0.547842               0.472673   \n",
       "ENSG00000243485.2               0.000000               0.000000   \n",
       "ENSG00000237613.2               0.000000               0.000000   \n",
       "ENSG00000268020.2               0.000000               0.000000   \n",
       "\n",
       "                   BT474_Persister_rep_3  \n",
       "ENSG00000223972.4               0.005309  \n",
       "ENSG00000227232.4               0.679085  \n",
       "ENSG00000243485.2               0.000000  \n",
       "ENSG00000237613.2               0.000000  \n",
       "ENSG00000268020.2               0.000000  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FPKM(gene_lengths=gene_lengths).set_output(transform=\"pandas\").fit_transform(counts.T).T.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8 scverse-lite",
   "language": "python",
   "name": "scverse-lite-py38"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
